### DISCUSSION TITLE
M4 Update Guide

### DISCUSSION THREAD

**Discussion by lloydtabb** [malloydata/malloy#1388] [Announcements]
# Updating your Malloy files to Malloy 4.0

With the Malloy 4.0 language changes, we're making Malloy more consistent readable and learnable.  You will first be getting warnings about these changes and ultimately, they will become errors.  Here is a list of changes you might encounter and ways to quickly fix them.

## the `table` function now takes a connection name.
Code that looks like:
```
source: carriers is table('duckdb:data/carriers.parquet') + {
  primary_key: code
  measure: carrier_count is count()
}
```
Becomes
```
source: carriers is duckdb.table('data/carriers.parquet') extend {
  primary_key: code
  measure: carrier_count is count()
}
```

Find and replace:
Find: `table\('([^']+):`
Replace `$1.table('`
![image](https://github.com/malloydata/malloy/assets/1093458/a099e696-04be-4fcc-84c4-e369a4b17a0b)

Then find all your sources and change '+' to 'extends
Find: `source: [^{]+`
![image](https://github.com/malloydata/malloy/assets/1093458/f01c1a84-836a-486c-a69e-83af616c74ce)

## count(distnct x) becomes count(x)
Find: `count(distinct`
Replace: `count(`

## `refine` is now `+`

this code:
```
  query: southwest_dashboard is carrier_dashboard refine {
    where: carriers.nickname ? 'Southwest'
  }
```
Becomes 
```
  query: southwest_dashboard is carrier_dashboard + {
    where: carriers.nickname ? 'Southwest'
  }
```
Find: ` refine `
Replace: ` + `
![image](https://github.com/malloydata/malloy/assets/1093458/79ae85de-92c4-46ae-ab4d-d8b9adeec431)

## In Sources, `query:` is now a `view:`

```
source: airports is duckdb.table('data/airports.parquet') + {
 query: by_state is {
   group_by: state
   aggregate: airport_count
  }
}
```
Is now:
```
source: airports is duckdb.table('data/airports.parquet') + {
 view: by_state is {
   group_by: state
   aggregate: airport_count
  }
}
```

Find:  `query: ([^-])+\{`
Replace: `view: $1{`

![image](https://github.com/malloydata/malloy/assets/1093458/d793cced-bebe-4340-a0ad-6a6216ae0c21)

**Comment by mtoy-googly-moogly** [comment]
These should all work ..

* **Replace:** `is ->` **with** `is `  (is space)
* **Replace** `query: ->` **with** `run:`
* **Replace** `project:` **with** `select:`
* **Replace** `count\(\*)` **with** `count()`
* **Replace** `count\(distinct\s+` **with** `count(`
* **Replace** `(source: .*)(\+)?\s{$` **with** `$1 extend {`
